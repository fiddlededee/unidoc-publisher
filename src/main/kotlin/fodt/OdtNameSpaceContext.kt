package fodt

import javax.xml.namespace.NamespaceContext

class OdtNameSpaceContext : NamespaceContext {
    override fun getNamespaceURI(p0: String?): String? {
        val namespaceURI = when (p0) {
            "text" -> "urn:oasis:names:tc:opendocument:xmlns:text:1.0"
            "style" -> "urn:oasis:names:tc:opendocument:xmlns:style:1.0"
            "fo" -> "urn:oasis:names:tc:opendocument:xmlns:xsl-fo-compatible:1.0"
            "office" -> "urn:oasis:names:tc:opendocument:xmlns:office:1.0"
            "table" -> "urn:oasis:names:tc:opendocument:xmlns:table:1.0"
            "draw" -> "urn:oasis:names:tc:opendocument:xmlns:drawing:1.0"
            "svg" -> "http://www.w3.org/2000/svg"
            "loext" -> "urn:org:documentfoundation:names:experimental:office:xmlns:loext:1.0"
            else -> null
        }
        return namespaceURI
    }

    override fun getPrefix(p0: String?): String {
        throw UnsupportedOperationException()
    }

    override fun getPrefixes(p0: String?): MutableIterator<String> {
        throw UnsupportedOperationException()
    }
}

