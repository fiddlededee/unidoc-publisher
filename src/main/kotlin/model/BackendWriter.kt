package model
interface BackendWriter {
    // pattern visitor
    fun write(openBlock: OpenBlock)
    fun write(image: Image)
    fun write(table: Table)
    fun write(tableRowGroup: TableRowGroup)
    fun write(tr: TableRow)
    fun write(td: TableCell)
    fun write(colGroup: ColGroup)
    fun write(col: Col)
    fun write(ol: OrderedList)
    fun write(ul: UnorderedList)
    fun write(li: ListItem)
    fun write(h: Heading)
    fun write(p: Paragraph)
    fun write(doc: Document)
    fun write(text: Text)
    fun write(span: Span)
    fun write(a: Anchor)
    fun write(toc: Toc)
    fun write(textFrame: TextFrame)
    fun write(footnote: Footnote)
    fun write(footnoteRef: FootnoteRef)
    fun write(dummyNode: Node)
}
