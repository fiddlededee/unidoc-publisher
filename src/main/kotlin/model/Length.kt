package model

class Length(var value: Float, var unit: LengthUnit = LengthUnit.perc) {
    companion object Factory {
        fun fromString(length: String?): Length? {
            if (length == null) return null
            return LengthUnit.entries
                .map {
                    val inValueUnit = when (it) {
                        LengthUnit.inch -> "in"; LengthUnit.perc -> "%"
                        LengthUnit.parrots -> ""; else -> it.toString()
                    }
                    """^([0-9]+[.]?[0-9]*)(${inValueUnit})$""".toRegex()
                        .matchEntire(length) to it
                }.filter { it.first != null }
                .getOrNull(0)
                ?.run { Length(first!!.groupValues[1].toFloat(), second) }
        }
    }
}

enum class LengthUnit { cm, mm, inch, px, pt, pc, em, ex, ch, rem, vw, vh, vmin, vmax, perc, parrots, mmm }
