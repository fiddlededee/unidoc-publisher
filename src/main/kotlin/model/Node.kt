package model

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.fasterxml.jackson.dataformat.yaml.YAMLGenerator

fun document(init: Document.() -> Unit = {}): Document {
    return Document().apply(init)
}

@JsonInclude(JsonInclude.Include.NON_EMPTY)
@JsonTypeInfo(include = JsonTypeInfo.As.WRAPPER_OBJECT, use = JsonTypeInfo.Id.NAME)
abstract class Node() {

    var id: String? = null
    var sourceTagName: String? = null
    var includeTags: MutableSet<String> = mutableSetOf()
    var sourceAttributes: MutableMap<String, String> = mutableMapOf()
    var sourceMapping: SourceMapping? = null

    @get:JsonIgnore
    abstract val isInline: Boolean

    @JsonIgnore
    private var parent: Node? = null

    private val children: ArrayList<Node> = arrayListOf()
    val roles = arrayListOf<String>()
    fun <T : Node> appendChild(childNode: T): T {
        val oldParent = childNode.parent
        children.add(childNode)
        childNode.parent = this
        oldParent?.removeChild(childNode)
        return childNode
    }

    fun <T : Node> appendFirstChild(childNode: T): T {
        if (children().size == 0)
            appendChild(childNode) else
            children()[0].insertBefore(childNode)
        return childNode
    }

    fun table(init: Table.() -> Unit = {}): Table {
        return appendChild(Table().apply(init))
    }

    fun img(src: String, init: Image.() -> Unit = {}): Image {
        return appendChild(Image(src).apply(init))
    }

    fun tableRowGroup(type: TRG, init: TableRowGroup.() -> Unit = {}): TableRowGroup {
        return appendChild(TableRowGroup(type).apply(init))
    }

    fun col(width: Length): Col {
        return appendChild(Col(width))
    }

    fun tr(init: TableRow.() -> Unit = {}): TableRow {
        return appendChild(TableRow().apply(init))
    }

    fun td(init: TableCell.() -> Unit = {}): TableCell {
        return appendChild(TableCell().apply(init))
    }

    fun h(level: Int, init: Heading.() -> Unit = {}): Heading {
        return appendChild(Heading(level).apply(init))
    }


    fun li(init: ListItem.() -> Unit = {}): ListItem {
        return appendChild(ListItem().apply(init))
    }

    fun ol(init: OrderedList.() -> Unit = {}): OrderedList {
        return appendChild(OrderedList().apply(init))
    }

    fun ul(init: UnorderedList.() -> Unit = {}): UnorderedList {
        return appendChild(UnorderedList().apply(init))
    }

    fun p(init: Paragraph.() -> Unit = {}): Paragraph {
        return appendChild(Paragraph().apply(init))
    }

    fun span(init: Span.() -> Unit = {}): Span {
        return appendChild(Span().apply(init))
    }

    fun a(href: String, init: Anchor.() -> Unit = {}): Anchor {
        return appendChild(Anchor(href).apply(init))
    }

    fun text(text: String): Text {
        return appendChild(Text(text))
    }

    fun openBlock(init: OpenBlock.() -> Unit = {}): OpenBlock {
        return appendChild(OpenBlock().apply(init))
    }

    operator fun String.unaryPlus(): Text {
        return this@Node.appendChild(Text(this))
    }

    private fun removeChild(childNode: Node) {
        children.remove(childNode)
    }

    fun remove() {
        val parentNode = parent
        if (parentNode != null) {
            parentNode.removeChild(this)
        } else throw Exception("Node doesn't hava a parent")
    }

    fun insertAfter(nodeToInsert: Node): Node {
        val oldParent = nodeToInsert.parent
        val parent = this.parent
            ?: throw Exception("Can't add node after another if it has no parent")
        val index = parent.children.indexOf(this)
        if (index == -1) {
            throw Exception("Didn't find object to add after")
        }
        parent.children.add(index + 1, nodeToInsert)
        nodeToInsert.parent = parent
        oldParent?.removeChild(nodeToInsert)
        return nodeToInsert
    }

    fun insertBefore(nodeToInsert: Node): Node {
        val oldParent = nodeToInsert.parent
        val parent = this.parent
            ?: throw Exception("Can't add node before another if it has no parent")
        val index = parent.children.indexOf(this)
        if (index == -1) {
            throw Exception("Didn't find object to add after")
        }
        parent.children.add(index, nodeToInsert)
        nodeToInsert.parent = parent
        oldParent?.removeChild(nodeToInsert)
        return nodeToInsert
    }


    // todo abstract ?
    abstract fun write(bw: BackendWriter)

    private fun ArrayList<Node>.fillWithDescendants(node: Node, filter: (Node) -> Boolean) {
        node.children.forEach {
            if (filter(it)) {
                add(it)
            }
            fillWithDescendants(it, filter)
        }
    }

    fun descendant(filter: (Node) -> Boolean = { true }): ArrayList<Node> {
        return arrayListOf<Node>().apply { fillWithDescendants(this@Node, filter) }
    }

    fun ancestor(filter: (Node) -> Boolean = { true }): ArrayList<Node> {
        val ancestors = arrayListOf<Node>()
        var parent = this@Node.parent()
        while (parent != null) {
            if (filter.invoke(parent)) ancestors.add(parent)
            parent = parent.parent()
        }
        return ancestors
    }

    fun nextSibling(filter: (Node) -> Boolean = { true }): ArrayList<Node> {
        val nextSibling = arrayListOf<Node>()
        val parent = this.parent() ?: return nextSibling
        if (parent.children.indexOf(this) == parent.children.size - 1) return nextSibling
        IntRange(parent.children.indexOf(this) + 1, parent.children.size - 1).forEach {
            if (filter(parent.children[it])) nextSibling.add(parent.children[it])
        }
        return nextSibling
    }

    fun previousSibling(filter: (Node) -> Boolean = { true }): ArrayList<Node> {
        val previousSibling = arrayListOf<Node>()
        val parent = this.parent() ?: return previousSibling
        if (parent.children.indexOf(this) == 0) return previousSibling
        IntRange(0, parent.children.indexOf(this) - 1).forEach {
            if (filter(parent.children[it])) previousSibling.add(parent.children[it])
        }
        return previousSibling
    }
    fun hasNext(): Boolean {
        val parent = this.parent() ?: throw Exception("The ${this::class.java.simpleName} has no parent")
        return (parent.children.indexOf(this) != parent.children.size - 1)
    }

    fun next(): Node {
        val parent = this.parent() ?: throw Exception("The ${this::class.java.simpleName} has no parent")
        if (parent.children.indexOf(this) == parent.children.size - 1)
            throw Exception("The ${this::class.java.simpleName} is last")
        return parent.children[parent.children.indexOf(this) + 1]
    }

    fun hasPrevious(): Boolean {
        val parent = this.parent() ?: throw Exception("The ${this::class.java.simpleName} has no parent")
        return (parent.children.indexOf(this) != 0)
    }

    fun previous(): Node {
        val parent = this.parent() ?: throw Exception("The ${this::class.java.simpleName} has no parent")
        if (parent.children.indexOf(this) == 0)
            throw Exception("The ${this::class.java.simpleName} is first")
        return parent.children[parent.children.indexOf(this) - 1]
    }

    fun index(): Int {
        val parent = this.parent()
            ?: throw Exception("The ${this::class.java.simpleName} node has no parent")
        return parent.children.indexOf(this)
    }

    @JsonProperty("children")
    fun getChildren(): ArrayList<Node> {
        return children
    }

    fun children(filter: (Node) -> Boolean = { true }): ArrayList<Node> {
        return arrayListOf<Node>()
            .apply { children.forEach { if (filter(it)) this.add(it) } }
    }

    fun parent(): Node? {
        return parent
    }

    fun roles(vararg role: String) {
        role.forEach { roles.add(it) }
    }

    fun toYamlString(): String {
        return ObjectMapper(YAMLFactory().disable(YAMLGenerator.Feature.WRITE_DOC_START_MARKER))
            .writeValueAsString(this)
    }

    fun <T : Node> T.wrapNodeInlineContents(): T {
        // Wrapping each inline nodes chains into paragraph node
        fun wrap(itemsToWrap: ArrayList<ArrayList<Node>>) {
            itemsToWrap.forEach { itemsChain ->
                if (itemsChain.isEmpty()) return@forEach
                itemsChain[0].insertBefore(Paragraph())
                    .apply { itemsChain.forEach { itemToWrap -> appendChild(itemToWrap) } }
            }
        }

        // Finding inline nodes chains
        val itemsToWrap = arrayListOf<ArrayList<Node>>()
        itemsToWrap.add(arrayListOf())
        this.children.forEach { nodeChild ->
            if (nodeChild.isInline) itemsToWrap.last().add(nodeChild)
            else if (itemsToWrap.last().isNotEmpty()) itemsToWrap.add(arrayListOf())
        }
        wrap(itemsToWrap)
        return this
    }


    // Don't understand html rendering rules, hope I'm right about this
    fun normalizeWhitespaces() {
        val localAst = this
        // Cleaning whitespaces after and before block nodes
        localAst.descendant { it is Text && it.text.isBlank() }.forEach { blankTextNode ->
            val localParent = blankTextNode.parent() ?: return@forEach
            val previousConsideredBlock =
                (!blankTextNode.hasPrevious() && !localParent.isInline) ||
                        (blankTextNode.hasPrevious() && (
                                !blankTextNode.previous().isInline ||
                                        blankTextNode.previous().sourceTagName == "br"
                                ))
            val nextConsideredBlock =
                (!blankTextNode.hasNext() && !localParent.isInline) ||
                        (blankTextNode.hasNext() && (
                                !blankTextNode.next().isInline ||
                                        blankTextNode.next().sourceTagName == "br"
                                ))

            if (previousConsideredBlock || nextConsideredBlock)
                blankTextNode.remove()
        }
        // Removing extra whitespaces
        Document().apply { appendChild(localAst) }.descendant { !it.isInline }.forEach { block ->
            if (block.sourceTagName == "pre" || block.ancestor { it.sourceTagName == "pre" }.isNotEmpty())
                return@forEach
            val nodesToNormalise = block.descendant { node ->
                (node is Image || node is Text || !node.isInline) &&
                        node.ancestor { !it.isInline }.firstOrNull() == block
            }
            var chain: ArrayList<Node> = arrayListOf()
            nodesToNormalise.forEach searchForChainsToNormalize@{ node ->
                if (!node.isInline) {
                    if (chain.isEmpty()) return@searchForChainsToNormalize else {
                        normalizeChain(chain)
                        chain = arrayListOf()
                    }
                } else chain.add(node)
            }
            normalizeChain(chain)
        }
    }

    private fun normalizeChain(chain: ArrayList<Node>) {
        if (chain.isEmpty()) return
        var trimStart = true
        chain.forEach byNodesToNormalize@{ node ->
            if (node is Image || !node.isInline) trimStart = false
            if (node !is Text) return@byNodesToNormalize
            if (trimStart) {
                node.text = node.text.trimStart()
            } else {
                if (node.text.trimStart() != node.text) node.text = " " + node.text.trimStart()
            }
            if (node.text.trimEnd() != node.text && node.text.isNotEmpty()) {
                node.text = node.text.trimEnd() + " "
                trimStart = true
            } else if (node.text.isNotEmpty()) trimStart = false
        }
        val iterator = chain.reversed().iterator()
        while (iterator.hasNext()) {
            val el = iterator.next()
            if (el !is Text) break
            if (el.text.isBlank()) el.remove() else {
                el.text = el.text.trimEnd()
                break
            }
        }

    }

    fun extractText(replaceRule: (Text) -> String = { it.text }): String {
        if (this is Text) return replaceRule.invoke(this)
        return descendant { it is Text }
            .map { it as Text }
            .joinToString("") { text -> replaceRule.invoke(text) }
    }

}

