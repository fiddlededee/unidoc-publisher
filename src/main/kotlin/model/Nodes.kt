package model

import java.util.*

class DummyNode() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

abstract class NoWriterNode() : Node() {
    override fun write(bw: BackendWriter) {
        throw Exception("Writer in NoWriterNodes should be defined in CustomStyle")
    }
}

class Document() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }

}

class OpenBlock() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class RectDimensions(val width: Length, val height: Length)

enum class AnchorType { ASCHAR }
class Image(
    var src: String,
    var width: Length? = null,
    var height: Length? = null,
    var anchorType: AnchorType = AnchorType.ASCHAR,
) : Node() {
    override val isInline: Boolean get() = true
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class Col(var width: Length) : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}


class ColGroup : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}


class Table() : Node() {
    // TODO: implement cols property
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

enum class TRG { head, body, foot }

class TableRowGroup(val type: TRG) : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class TableRow() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class TableCell(var colspan: Int = 1, var rowspan: Int = 1) : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }

//    fun wrapNodeInlineContents(): TableCell {
//        // Wrapping each inline nodes chains into paragraph node
//        fun wrap(itemsToWrap: ArrayList<ArrayList<Node>>) {
//            itemsToWrap.forEach { itemsChain ->
//                if (itemsChain.isEmpty()) return@forEach
//                itemsChain[0].addBefore(Paragraph())
//                    .apply { itemsChain.forEach { itemToWrap -> addChild(itemToWrap) } }
//            }
//        }
//
//        // Finding inline nodes chains
//        val itemsToWrap = arrayListOf<ArrayList<Node>>()
//        itemsToWrap.add(arrayListOf())
//        this.children.forEach { nodeChild ->
//            if (nodeChild.isInline) itemsToWrap.last().add(nodeChild)
//            else if (itemsToWrap.last().isNotEmpty()) itemsToWrap.add(arrayListOf())
//        }
//        wrap(itemsToWrap)
//        return this
//    }
}

class OrderedList() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class UnorderedList() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class ListItem() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }

    fun wrapListItemContents(): ListItem {
        // Wrapping each inline nodes chains into paragraph node
        fun wrap(itemsToWrap: ArrayList<ArrayList<Node>>) {
            itemsToWrap.forEach { itemsChain ->
                if (itemsChain.isEmpty()) return@forEach
                itemsChain[0].insertBefore(Paragraph())
                    .apply { itemsChain.forEach { itemToWrap -> appendChild(itemToWrap) } }
            }
        }

        // Finding inline nodes chains
        val itemsToWrap = arrayListOf<ArrayList<Node>>()
        itemsToWrap.add(arrayListOf())
        this.children().forEach { listItemChild ->
            if (listItemChild.isInline) itemsToWrap.last().add(listItemChild)
            else if (itemsToWrap.last().isNotEmpty()) itemsToWrap.add(arrayListOf())
        }
        wrap(itemsToWrap)
        return this
    }
}

class Header(var level: Int) : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class Paragraph() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

// Look like excessive
//class Pre() : Node() {
//    override val isInline: Boolean get() = false
//    override fun write(bw: BackendWriter) {
//        bw.write(this)
//    }
//}
//
//class Code() : Node() {
//    override val isInline: Boolean get() = false
//    override fun write(bw: BackendWriter) {
//        bw.write(this)
//    }
//}


class Span() : Node() {
    override val isInline: Boolean get() = true
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class Anchor(var href: String) : Node() {
    override val isInline: Boolean get() = true
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class Text(var text: String) : Node() {
    override val isInline: Boolean get() = true
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class Toc(var levels: Int = 3, title: String = "Table of contents 111") : Node() {
    var titleNode = Paragraph().apply { +title }
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class TextFrame() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class Footnote(
    val footnoteId: String = UUID.randomUUID().toString()
) : Node() {
    override val isInline: Boolean get() = true
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
    init {
        id = footnoteId
    }
}

class FootnoteRef(val footnote: Footnote) : Node() {
    override val isInline: Boolean get() = true
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}
