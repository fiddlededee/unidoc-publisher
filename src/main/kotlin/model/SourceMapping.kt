package model

class SourceMapping (var Lineno : Int, var file : String?)