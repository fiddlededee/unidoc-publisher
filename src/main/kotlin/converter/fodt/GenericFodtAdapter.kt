package converter.fodt

import model.Node
import reader.HtmlNode
import reader.UnknownTagProcessing
import writer.OdtStyleList

interface GenericFodtAdapter {
    val unknownTagProcessingRule: HtmlNode.() -> UnknownTagProcessing
    fun basicStyle(): OdtStyleList
    fun Node.normalizeAll() {}
}