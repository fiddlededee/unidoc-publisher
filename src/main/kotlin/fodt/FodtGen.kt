package fodt

import converter.PreRegion
import org.w3c.dom.Document
import org.w3c.dom.Element
import org.w3c.dom.Node
import org.w3c.dom.NodeList
import org.xml.sax.InputSource
import java.io.StringReader
import java.io.StringWriter
import java.util.NoSuchElementException
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.transform.OutputKeys
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult
import javax.xml.xpath.XPath
import javax.xml.xpath.XPathConstants
import javax.xml.xpath.XPathFactory

val styleNS = "urn:oasis:names:tc:opendocument:xmlns:style:1.0"
val foNS = "urn:oasis:names:tc:opendocument:xmlns:xsl-fo-compatible:1.0"

val xpathInstance: XPath = XPathFactory.newInstance().newXPath().apply {
    namespaceContext = OdtNameSpaceContext()
}


fun Document.serialize(): String {
    val transformerFactory = TransformerFactory.newInstance()
    val trans = transformerFactory.newTransformer()
    trans.setOutputProperty(OutputKeys.METHOD, "xml")
    val sw = StringWriter()
    val result = StreamResult(sw)
    val source = DOMSource(this.documentElement)
    trans.transform(source, result)
    return sw.toString()
}


fun String.parseStringAsXML(): Document {
    return DocumentBuilderFactory.newInstance()
        .apply { isNamespaceAware = true }
        .newDocumentBuilder()
        .parse(InputSource(StringReader(this)))
}


fun Node.xpath(xpath: String): NodeList {
    return xpathInstance.compile(xpath)
        .evaluate(this, XPathConstants.NODESET) as NodeList
}


fun NodeList.iterable(): Iterable<Node> = object : Iterable<Node> {
    override fun iterator(): Iterator<Node> = object : Iterator<Node> {
        private var index = 0
        override fun hasNext(): Boolean = index < this@iterable.length
        override fun next(): Node {
            if (!hasNext()) throw NoSuchElementException()
            return this@iterable.item(index++)
        }
    }
}

class FodtGenerator(val preFodtList: ArrayList<PreRegion>, template: String) {
    val enrichedTemplate = template.parseStringAsXML()
    private val automaticStyles: Document = DocumentBuilderFactory.newInstance().apply { isNamespaceAware = true }
        .newDocumentBuilder().parse(InputSource(StringReader("<root/>")))


    fun serialize(): String {
        return enrichedTemplate.serialize()
    }

    enum class StyleFamily { PARAGRAPH, TEXT, TABLE, TABLECELL, TABLECOLUMN, TABLEROW, GRAPHIC }

    var styleNum = 0
    fun setStyles(xpathExpression: String, styleFamily: StyleFamily, preFodt: Document) {
        val nodesWithStyles = preFodt.xpath(xpathExpression)
        nodesWithStyles.iterable().forEach {
            val styleNamePrefix = when (styleFamily) {
                StyleFamily.PARAGRAPH -> "PAMC"
                StyleFamily.TEXT -> "TAMC"
                StyleFamily.TABLE -> "TABAMC"
                StyleFamily.TABLECELL -> "TABCELLAMC"
                StyleFamily.TABLECOLUMN -> "TABCOLAMC"
                StyleFamily.TABLEROW -> "TABROWAMC"
                StyleFamily.GRAPHIC -> "GAMC"
            }
            val styleFamilyName = when (styleFamily) {
                StyleFamily.PARAGRAPH -> "paragraph"
                StyleFamily.TEXT -> "text"
                StyleFamily.TABLE -> "table"
                StyleFamily.TABLECELL -> "table-cell"
                StyleFamily.TABLECOLUMN -> "table-column"
                StyleFamily.TABLEROW -> "table-row"
                StyleFamily.GRAPHIC -> "graphic"
            }
            styleNum += 1
            val generatedStyleName = "$styleNamePrefix$styleNum"
            val automaticStylesStyle =
                automaticStyles.createElementNS(styleNS, "style:style")
            automaticStylesStyle.apply {
                setAttributeNS(styleNS, "style:name", generatedStyleName)
                setAttributeNS(styleNS, "style:family", styleFamilyName)
            }
            automaticStyles.documentElement.appendChild(automaticStylesStyle)
            val textPropertiesNodes = it.childNodes.iterable().filter { node ->
                node is Element && node.nodeName == "style:text-properties"
            }
            if (textPropertiesNodes.isNotEmpty()) {
                val textPropertiesNode = textPropertiesNodes[0]
                automaticStyles.adoptNode(textPropertiesNode)
                automaticStylesStyle.appendChild(textPropertiesNode)
            }
            val paragraphPropertiesNodes = it.childNodes.iterable().filter { node ->
                node is Element && node.nodeName == "style:paragraph-properties"
            }
            if (paragraphPropertiesNodes.isNotEmpty()) {
                val paragraphPropertiesNode = paragraphPropertiesNodes[0]
                automaticStyles.adoptNode(paragraphPropertiesNode)
                automaticStylesStyle.appendChild(paragraphPropertiesNode)
            }
            val graphicPropertiesNodes = it.childNodes.iterable().filter { node ->
                node is Element && node.nodeName == "style:graphic-properties"
            }
            if (graphicPropertiesNodes.isNotEmpty()) {
                val graphicPropertiesNode = graphicPropertiesNodes[0]
                automaticStyles.adoptNode(graphicPropertiesNode)
                automaticStylesStyle.appendChild(graphicPropertiesNode)
            }
            val tablePropertiesNodes = it.childNodes.iterable().filter { node ->
                node is Element && (node.nodeName == "style:table-properties"
                        || node.nodeName == "style:table-column-properties"
                        || node.nodeName == "style:table-cell-properties"
                        || node.nodeName == "style:table-row-properties")
            }
            if (tablePropertiesNodes.isNotEmpty()) {
                val tablePropertiesNode = tablePropertiesNodes[0]
                automaticStyles.adoptNode(tablePropertiesNode)
                automaticStylesStyle.appendChild(tablePropertiesNode)
            }
            val styleMasterPageName = (it as Element).getAttributeNode("style:master-page-name")
            if (styleMasterPageName != null) {
                automaticStylesStyle.apply {
                    setAttributeNS(
                        styleNS,
                        "style:master-page-name",
                        styleMasterPageName.nodeValue.replace(" ", "_20_")
                    )
                }
            }
            if (arrayOf(StyleFamily.TABLE, StyleFamily.TABLECELL, StyleFamily.TABLECOLUMN, StyleFamily.TABLEROW)
                    .contains(styleFamily)
            ) {
                it.setAttribute("table:style-name", generatedStyleName)
            } else {
                val styleNameAttribute =
                    if (styleFamily == StyleFamily.GRAPHIC)
                        it.getAttributeNode("draw:style-name") else
                        it.getAttributeNode("text:style-name")
                if (styleNameAttribute != null) {
                    automaticStylesStyle.apply {
                        setAttributeNS(
                            styleNS,
                            "style:parent-style-name",
                            styleNameAttribute.nodeValue.replace(" ", "_20_")
                        )
                    }
                    styleNameAttribute.nodeValue = generatedStyleName
                } else {
                    if (styleFamily == StyleFamily.GRAPHIC)
                        it.setAttribute("draw:style-name", generatedStyleName) else
                        it.setAttribute("text:style-name", generatedStyleName)
                }
            }
        }
    }

    fun processTaggedRegion(preRegion: PreRegion) {
        val preFodt = preRegion.pre.parseStringAsXML()
        setStyles("(//text:p|//text:h)", StyleFamily.PARAGRAPH, preFodt)
        setStyles("(//text:span)", StyleFamily.TEXT, preFodt)
        setStyles("(//table:table)", StyleFamily.TABLE, preFodt)
        setStyles("(//table:table-cell)", StyleFamily.TABLECELL, preFodt)
        setStyles("(//table:table-column)", StyleFamily.TABLECOLUMN, preFodt)
        setStyles("(//table:table-row)", StyleFamily.TABLEROW, preFodt)
        setStyles("(//draw:frame)", StyleFamily.GRAPHIC, preFodt)
        val officeText = enrichedTemplate.xpath("//office:body/office:text")
        if (officeText.length == 1) {
            val officeTextElement = officeText.item(0)
            officeTextElement.xpath("//text:variable-set[@text:name = 'include']")
                .iterable().map { it as Element }.forEach { markedElement ->
                    val tagName = markedElement.textContent
                    if (!preRegion.includeTags.contains(tagName)) return@forEach
                    val elementToReplace = if (preRegion.inLine) markedElement else
                        markedElement.parentNode
                    preFodt.xpath("/root/*").iterable().forEach {
                        enrichedTemplate.adoptNode(it)
                        elementToReplace.parentNode.insertBefore(it, elementToReplace)
                    }
                    elementToReplace.parentNode.removeChild(elementToReplace)
                }
            officeTextElement.xpath("//text:variable-set[@text:name = 'process']")
                .iterable().map { it as Element }.forEach {
                    if (it.textContent != "end") return@forEach
                    var deleteFromNode = it.parentNode
                    while (deleteFromNode.nextSibling != null) {
                        val nextSibling = deleteFromNode.nextSibling
                        deleteFromNode.parentNode.removeChild(deleteFromNode)
                        deleteFromNode = nextSibling
                    }
                    deleteFromNode.parentNode.removeChild(deleteFromNode)
                }
        } else {
            throw Exception("office:body/office:text not found")
        }

    }

    init {
        preFodtList.forEach { preRegion ->
            processTaggedRegion(preRegion)
        }

        val officeAutomaticStyles = enrichedTemplate.xpath("//office:automatic-styles")
        if (officeAutomaticStyles.length == 1) {
            val officeAutomaticStylesElement = officeAutomaticStyles.item(0)
            automaticStyles.xpath("/root/*").iterable().forEach {
                enrichedTemplate.adoptNode(it)
                officeAutomaticStylesElement.appendChild(it)
            }
        } else {
            throw Exception("office:automatic-styles not found")
        }
    }

}

