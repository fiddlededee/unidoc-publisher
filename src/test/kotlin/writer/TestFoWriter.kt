package writer

import fodt.iterable
import fodt.parseStringAsXML
import fodt.serialize
import fodt.xpath
import model.*
import org.apache.fop.apps.FopFactory
import org.apache.fop.apps.MimeConstants
import org.approvaltests.Approvals
import org.junit.jupiter.api.Test
import java.io.File
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.sax.SAXResult
import org.apache.fop.apps.Fop
import org.redundent.kotlin.xml.PrintOptions
import org.w3c.dom.Element
import java.io.FileOutputStream
import javax.xml.transform.Transformer
import javax.xml.transform.Source
import org.w3c.dom.Document as DomDocument
import javax.xml.transform.Result


class TestFoWriter {

    @Test
    fun simple() {
        val doc = Document().apply {
            p { +"Some paragraph" }
        }
        val foStyleList = FoStyleList()
        FoWriter(foStyleList = foStyleList)
            .apply { doc.write(this) }
            .preFoNode
            .toString()
            .apply { Approvals.verify(this) }
    }
}