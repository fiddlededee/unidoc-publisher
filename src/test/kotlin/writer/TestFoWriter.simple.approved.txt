<fo:root xmlns:fo="http://www.w3.org/1999/XSL/Format">
	<fo:block>
		Some paragraph
	</fo:block>
</fo:root>